lib/vm/src/instance.rs

```diff
@@ -761,110 +661,272 @@ impl Instance {
 
-        let layout = Instance::alloc_layout(&offsets);
+        Layout::from_size_align(size, align).unwrap()
+    }
+
+    /// Allocate `Instance` (it is an uninitialized pointer).
+    ///
+    /// `offsets` is used to compute the layout with `Self::instance_layout`.
+    fn allocate_instance(offsets: &VMOffsets) -> (NonNull<Instance>, Layout) {
+        let layout = Self::instance_layout(offsets);
 
         #[allow(clippy::cast_ptr_alignment)]
         let instance_ptr = unsafe { alloc::alloc(layout) as *mut Instance };
+
         let ptr = if let Some(ptr) = NonNull::new(instance_ptr) {
-            ptr.cast()
+            ptr
         } else {
             alloc::handle_alloc_error(layout);
         };
 
-        (ptr, offsets)
+        (ptr, layout)
     }
 
-    /// Get the locations of where the local `VMMemoryDefinition`s should be stored.
-    ///
-    /// This function lets us create `Memory` objects on the host with backing
-    /// memory in the VM.
+    /// Deallocate `Instance`.
     ///
     /// # Safety
-    /// - `instance_ptr` must point to enough memory that all of the offsets in
-    ///   `offsets` point to valid locations in memory.
-    pub unsafe fn memory_definition_locations(
-        instance_ptr: NonNull<u8>,
-        offsets: &VMOffsets,
-    ) -> Vec<NonNull<VMMemoryDefinition>> {
-        let num_memories = offsets.num_local_memories;
-        let num_memories = usize::try_from(num_memories).unwrap();
-        let mut out = Vec::with_capacity(num_memories);
-        // TODO: better encapsulate this logic, this shouldn't be duplicated
-        let base_ptr = instance_ptr.as_ptr().add(std::mem::size_of::<Instance>());
-        for i in 0..num_memories {
-            let mem_offset = offsets.vmctx_vmmemory_definition(LocalMemoryIndex::new(i));
-            let mem_offset = usize::try_from(mem_offset).unwrap();
+    ///
+    /// `Self.instance` must be correctly set and filled before being
+    /// dropped and deallocated.
+    unsafe fn deallocate_instance(&mut self) {
+        let instance_ptr = self.instance.as_ptr();
 
-            let new_ptr = NonNull::new_unchecked(base_ptr.add(mem_offset));
+        ptr::drop_in_place(instance_ptr);
+        std::alloc::dealloc(instance_ptr as *mut u8, self.instance_layout);
+    }
 
-            out.push(new_ptr.cast());
-        }
-        out
+    /// Get the number of strong references pointing to this
+    /// `InstanceAllocator`.
+    pub fn strong_count(&self) -> usize {
+        self.strong.load(atomic::Ordering::SeqCst)
     }
 
-    /// Get the locations of where the `VMTableDefinition`s should be stored.
+    /// Get a reference to the `Instance`.
+    #[inline]
+    pub(crate) fn instance_ref<'a>(&'a self) -> &'a Instance {
+        // SAFETY: The pointer is properly aligned, it is
+        // “dereferencable”, it points to an initialized memory of
+        // `Instance`, and the reference has the lifetime `'a`.
+        unsafe { self.instance.as_ref() }
+    }
```

Comment:
`self.instance` is a correctly aligned and initialized pointer to `Instance`  in `new`, and in `instance_ref` (now renamed `as_ref`). I don't see why this method or `new` must be `unsafe`?

Reply:
Also, the pointer to `Instance` is only owned, and can only be freed by `InstanceAllocator`. Only `&Instance` can be used outside this struct. So except if someone deref the reference and do something bad with it, this code above is safe :-p.

Reply:
> self.instance is a correctly aligned and initialized pointer to Instance in new, and in instance_ref (now renamed as_ref). I don't see why this method or new must be unsafe?

I don't believe we can guarantee that with safe Rust, therefore we can't make that assumption without `unsafe`. `unsafe` means "read the docs for how to use this function correctly", if there's a way to use the function that compiles without `unsafe` that will cause memory safety issues, it's unsound and the API should be marked `unsafe`.

Reply:
Notably, anything that takes a `NonNull<T>` that assumes the pointer is anything but garbage must be `unsafe` because `NonNull::dangling` can be used in safe Rust to get a pointer to arbitrary memory.

Reply:
Because this isn't a public method, I think this is a relatively small deal, but I think it's still best to be precise here!

Reply:
I see! If you consider `InstanceAllocator` alone, then yes, I fully agree. I was talking about `InstanceHandle` + `InstanceAllocator`, that's why I wasn't aligned with you! I'm updating the code now to make `InstanceAllocator::new` unsafe. If the constructor is safely called, then `as_ref` will be safe too.

Done!